pkg/ingester-rf1/metastore/metastore_hack.go

```diff
@@ -60,10 +66,10 @@ func (m *metastoreState) applyTruncate(request *raftlogpb.TruncateCommand) (*any
 		_ = level.Error(m.logger).Log("msg", "failed to get metadata bucket", "err", err)
 		return nil, err
 	}
-	for k, segment := range m.segments {
-		if ulid.MustParse(segment.Id).Time() < request.Timestamp {
-			if err = bucket.Delete([]byte(segment.Id)); err != nil {
-				_ = level.Error(m.logger).Log("msg", "failed to delete stale segments", "err", err)
+	for k, v := range m.segments {
+		if ulid.MustParse(v.Id).Time() < request.Timestamp {
+			if err = m.deleteSegment(context.TODO(), bucket, v); err != nil {
```

Comment:
I'm wondering if we should delete from objstore async?
We do the the whole applyTruncate operation under the state lock, and deleting several hundred files from object storage will take a few seconds at least, during which we can't accept writes

Reply:
Do we need to hold the state lock for the entire `applyTruncate` function? BoltDB is thread-safe and we are using transactions.

Reply:
You're right - I think we just need the lock for the local state map update but if we only lock for that then we run into the same problem of local state being different to the actual data in storage for a short period of time